File: python/middleOfLinkedList#876/try1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cur = head
        n = 0
        while(cur):
            cur = cur.next
            n += 1

        cur = head
        for i in range(n//2):
            cur = cur.next

        return cur
        

# Time: O(n)
# Space: O(1)

File: c++/middleOfLinkedList#876/try1.cpp
// this is a poor old solution from when I first started

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) 
    {
        ListNode* curr = head;
        int count = 0;
        while(curr)
        {
            count++;
            curr = curr->next;
        }   
        int mid = (count/2) + 1;

        int newCount = 1;
        ListNode* newCurr = head;
        while(newCount < mid)
        {
            newCount++;
            newCurr = newCurr->next;
        }
        return newCurr;
    }
};
